fix(locator): improve error message when unchecking radio buttons

dev-shahed opened the pull request:
fix(locator): improve error message when unchecking radio buttons

This patch improves the error message when attempting to uncheck radio buttons using uncheck(), setChecked(false), or page.uncheck(). Previously, these operations would fail with a confusing error message "Clicking the checkbox did not change its state" that didn't clearly explain why the operation failed.

Radio buttons cannot be unchecked by clicking - they can only be unchecked by selecting another radio button in the same group. The new error message clearly explains this behavior and guides users toward the correct approach.

**Before:** Error: locator.uncheck: Clicking the checkbox did not change its state

**After:** Error: locator.uncheck: Cannot uncheck radio button. Radio buttons can only be unchecked by selecting another radio button in the same group.


**Changes:**
- Added `isRadioButton()` utility method to injected script for radio button detection
- Enhanced `_setChecked()` method in ElementHandle to detect radio buttons and provide clear error messaging
- Added test case to verify the new error behavior

**Testing:**
- All existing check/uncheck tests continue to pass
- New test verifies proper error message for radio button uncheck attempts
- Checkbox functionality remains unaffected

Fixes #37178

dev-shahed committed:
feat(dom): add isRadioButton method and enforce radio button uncheck rules

dev-shahed committed:
fix(dom): refine radio button uncheck logic and clean up comment

dgozman commented on packages/playwright-core/src/server/dom.ts:
```diff
@@ -732,9 +732,22 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
     };
+
+    const isRadio = async () => {
+      const result = await progress.race(this.evaluateInUtility(([injected, node]) => injected.isRadioButton(node), {}));
```
Could you please merge this evaluate into `isChecked()`?

dev-shahed replied:
Yes, done! I merged the radio button detection into the existing elementState('checked') call instead of making a separate evaluation. Now it returns both the checked state and radio button type in a single browser call, which is more efficient. Thanks..

dgozman reviewed (commented):
Thank you for the PR! This looks good overall, but bots are unhappy. Could you please address the linter and my comment?

dev-shahed committed:
fix(locator): improve error message when unchecking radio buttons

This patch improves the error message when attempting to uncheck radio buttons using uncheck(), setChecked(false), or page.uncheck(). Previously, these operations would fail with a confusing error message "Clicking the checkbox did not change its state" that didn't clearly explain why the operation failed.

Radio buttons cannot be unchecked by clicking - they can only be unchecked by selecting another radio button in the same group. The new error message clearly explains this behavior and guides users toward the correct approach.

Changes:
- Enhanced ElementStateQueryResult type to include isRadio property
- Modified elementState method to detect radio buttons when checking state
- Added radio button detection in _setChecked method to throw clear error

Before:
Error: locator.uncheck: Clicking the checkbox did not change its state

After:
Error: locator.uncheck: Cannot uncheck radio button. Radio buttons can only be unchecked by selecting another radio button in the same group.

dev-shahed committed:
refactor(injected): remove unnecessary whitespace in injectedScript.ts

dev-shahed commented:
@microsoft-github-policy-service agree

dgozman reviewed (approved):
Looks great, merging in. Thank you for the PR!
